modules/hal_nordic/nrfs/backends/nrfs_backend_ipc_service.c

```diff
@@ -0,0 +1,237 @@
+	.cb = {
+		.bound    = ipc_sysctrl_ept_bound,
+		.received = ipc_sysctrl_ept_recv,
+	},
+};
+
+static struct ipc_channel_config_t ipc_cpusys_channel_config = {
+	.ipc_instance	 = DEVICE_DT_GET(DT_ALIAS(ipc_to_cpusys)),
+	.endpoint_config = &ipc_sysctrl_ept_cfg,
+	.status		 = ATOMIC_INIT(NOT_CONNECTED),
+	.enabled	 = true
+};
+
+/**
+ * @brief nrfs backend error handler
+ *
+ * @param error_id The id of an error to handle.
+ * @param error additional error code if needed, if not needed use 0.
+ * @param fatal true if fatal error and needs special handling
+ */
+__weak void nrfs_backend_error_handler(enum nrfs_backend_error error_id, int error, bool fatal)
+{
+	switch (error_id) {
+	case NRFS_ERROR_EPT_RECEIVE_DATA_TOO_LONG:
+		LOG_ERR("Received data is too long. Config error.");
+		break;
+
+	case NRFS_ERROR_NO_DATA_RECEIVED:
+		LOG_ERR("No data in received message!");
+		break;
+
+	case NRFS_ERROR_IPC_OPEN_INSTANCE:
+		LOG_ERR("IPC open instance failure with error: %d", error);
+		break;
+
+	case NRFS_ERROR_IPC_REGISTER_ENDPOINT:
+		LOG_ERR("IPC register endpoint failure with error: %d", error);
+		break;
+
+	default:
+		LOG_ERR("Undefined error id: %d, error cause: %d", error_id, error);
+		break;
+	}
+
+	if (fatal) {
+		nrfs_backend_fatal_error_handler(error_id);
+	}
+}
+
+static void ipc_sysctrl_ept_bound(void *priv)
+{
+	LOG_INF("Bound to sysctrl.");
+	k_event_post(&ipc_connected_event, IPC_INIT_DONE_EVENT);
+	atomic_set(&ipc_cpusys_channel_config.status, CONNECTED);
+}
+
+static void ipc_sysctrl_ept_recv(const void *data, size_t size, void *priv)
+{
+	__ASSERT(size <= MAX_PACKET_DATA_SIZE, "Received data is too long. Config error.");
+	if (size <= MAX_PACKET_DATA_SIZE) {
+		rx_data.channel_id = IPC_CPUSYS_CHANNEL_ID;
+		rx_data.size	   = size;
+		if (data) {
+			memcpy(rx_data.data, (uint8_t *)data, size);
+			nrfs_dispatcher_notify(&rx_data.data, rx_data.size);
+		} else {
+			nrfs_backend_error_handler(NRFS_ERROR_NO_DATA_RECEIVED, 0, false);
+		}
+	} else {
+		nrfs_backend_error_handler(NRFS_ERROR_EPT_RECEIVE_DATA_TOO_LONG, 0, true);
+	}
+}
+
+static void nrfs_backend_send_work(struct k_work *item)
+{
+	LOG_DBG("Sending data from workqueue");
+	static struct ipc_data_packet_t data_to_send;
+
+	while (k_msgq_get(&ipc_transmit_msgq, &data_to_send, K_NO_WAIT) == 0) {
+		ipc_service_send(&ipc_cpusys_channel_config.ipc_ept, &tx_data.data, tx_data.size);
+	}
+}
+
+/**
+ *  @brief Initialize ipc channel
+ *
+ *  @param ipc_channel_config pointer to ipc channel config structure
+ *
+ *  @return -EINVAL when instance configuration is invalid.
+ *  @return -EIO when no backend is registered.
+ *  @return -EALREADY when the instance is already opened (or being opened).
+ *  @return -EBUSY when the instance is busy.
+ *  @return 0 on suscess
+ */
+static int ipc_channel_init(void)
+{
+	int ret = 0;
+
+	k_work_init(&backend_send_work, nrfs_backend_send_work);
+
+	struct ipc_channel_config_t *ipc_channel_config = &ipc_cpusys_channel_config;
```

Comment:
No reason to avoid concise variable names, please place the variable definition at the top of a block.
```c
	struct ipc_channel_config_t *ch_cfg;
	int ret = 0;

	k_work_init(&backend_send_work, nrfs_backend_send_work);
	ch_cfg = &ipc_cpusys_channel_config;
```